src/thread.rs

```diff
@@ -464,6 +673,68 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
         return free_tls_statics;
     }
 
+    /// Called when the current thread performed
+    /// an atomic operation that potentially makes
+    /// progress
+    fn thread_yield_progress(&mut self) {
+        log::trace!("Thread {:?} has made progress via generic progress", self.active_thread);
+        self.threads[self.active_thread].yield_state.on_progress();
+    }
+
+    /// Called when the current thread performs
+    /// an atomic operation that may be visible to other threads.
+    fn thread_yield_atomic_wake(&mut self, alloc_id: AllocId, offset: Size, len: Size) {
+        let threads = &mut self.threads;
+
+        // This thread performed an atomic update, mark as making progress.
+        log::trace!("Thread {:?} has made progress via atomic store", self.active_thread);
+        threads[self.active_thread].yield_state.on_progress();
+
+        // Awake all threads that were awaiting on changes to the modified atomic.
+        self.yielding_thread_set.drain_filter(move |&thread_id| {
+            let thread = &mut threads[thread_id];
+            if thread.yield_state.should_wake_atomic(alloc_id, offset, len) {
+                thread.state = ThreadState::Enabled;
```

Comment:
A thread that's currently recording might or might not be yield-blocked, right? So this could be enabling a thread that isn't yield-blocked? Couldn't this be a problem if the thread is currently blocked on something else, like a mutex? I am not sure if all our scheduling primitives are robust wrt. spurious wakeups...